src/Zio/FileSystems/ZipArchiveFileSystem.cs

```diff
@@ -0,0 +1,1034 @@
+                if (fileData != null)
+                {
+                    // we only check for read share, because ZipArchive doesn't support write share
+                    if (share is not FileShare.Read and not FileShare.ReadWrite)
+                    {
+                        throw new IOException("File is already opened for reading");
+                    }
+
+                    if (fileShare is not FileShare.Read and not FileShare.ReadWrite)
+                    {
+                        throw new IOException("File is already opened for reading by another stream with non compatible share");
+                    }
+
+                    fileData.Second++;
+                }
+                else
+                {
+                    _fileSystem._openStreams.Add(_entry, new Pair<FileShare, int>(share, 1));
+                }
+
+                Share = share;
+                _streamImplementation = entry.Open();
+            }
+
+            private FileShare Share { get; }
+
+            public override bool CanRead => _streamImplementation.CanRead;
+
+            public override bool CanSeek => _streamImplementation.CanSeek;
+
+            public override bool CanWrite => _streamImplementation.CanWrite;
+
+            public override long Length => _streamImplementation.Length;
+
+            public override long Position
+            {
+                get => _streamImplementation.Position;
+                set => _streamImplementation.Position = value;
+            }
+
+            public override void Flush()
+            {
+                _streamImplementation.Flush();
+            }
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                return _streamImplementation.Read(buffer, offset, count);
+            }
+
+            public override long Seek(long offset, SeekOrigin origin)
+            {
+                return _streamImplementation.Seek(offset, origin);
+            }
+
+            public override void SetLength(long value)
+            {
+                _streamImplementation.SetLength(value);
+            }
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                _streamImplementation.Write(buffer, offset, count);
+            }
+
+            public override void Close()
+            {
+                if (_isDisposed)
+                {
+                    return;
+                }
+
+                _streamImplementation.Close();
+                _isDisposed = true;
+                _fileSystem._openStreams.TryGetValue(_entry, out var fileData);
+                fileData.Second--;
+                if (fileData.Second == 0)
+                {
+                    _fileSystem._openStreams.Remove(_entry);
+                }
+            }
+        }
+
+        private class Pair<T, U>
```

Comment:
Does it have to be a generic type if its only internal usage is `Pair<FileShare, int>`?

Reply:
Maybe rename to `EntryState` or something similar?